test/suite/test_compact12.py

```diff
@@ -0,0 +1,143 @@
+# This test creates:
+#
+# - One table with the first 1/4 of keys deleted and obsolete and the last 10th of the keys
+#   removed with fast truncate but not yet obsolete.
+#
+# It checks that:
+#
+# - Compaction correctly rewrites pages in WT_REF_DELETED state but are still on disk.
+class test_compact12(wttest.WiredTigerTestCase):
+    create_params = 'key_format=i,value_format=S,allocation_size=4KB,leaf_page_max=32KB,leaf_value_max=16MB'
+    conn_config = 'cache_size=100MB,statistics=(all),verbose=[compact:4]'
+    uri_prefix = 'table:test_compact12'
+
+    table_numkv = 10 * 1000
+    value_size = kilobyte # The value should be small enough so that we don't create overflow pages.
+
+    def get_size(self, uri):
+        stat_cursor = self.session.open_cursor('statistics:' + uri, None, 'statistics=(all)')
+        size = stat_cursor[stat.dsrc.block_size][2]
+        stat_cursor.close()
+        return size
+
+    def get_fast_truncated_pages(self):
+        stat_cursor = self.session.open_cursor('statistics:', None, None)
+        pages = stat_cursor[stat.conn.rec_page_delete_fast][2]
+        stat_cursor.close()
+        return pages
+
+    def truncate(self, uri, key1, key2):
+        lo_cursor = self.session.open_cursor(uri)
+        hi_cursor = self.session.open_cursor(uri)
+        lo_cursor.set_key(key1)
+        hi_cursor.set_key(key2)
+        self.session.truncate(None, lo_cursor, hi_cursor, None)
+        lo_cursor.close()
+        hi_cursor.close()
+
+    def populate(self, uri, num_keys):
+        c = self.session.open_cursor(uri, None)
+        for k in range(num_keys // 10 * 9):
+            self.session.begin_transaction()
+            c[k] = ('%07d' % k) + '_' + 'a' * self.value_size
+            self.session.commit_transaction(f'commit_timestamp={self.timestamp_str(2)}')
+
+        for k in range(num_keys // 10 * 9, num_keys):
+            self.session.begin_transaction()
+            c[k] = ('%07d' % k) + '_' + 'b' * self.value_size
+            self.session.commit_transaction(f'commit_timestamp={self.timestamp_str(2)}')
+        c.close()
+
+    def test_compact12_truncate(self):
+        # FIXME-WT-11399
+        if self.runningHook('tiered'):
+            self.skipTest("this test does not yet work with tiered storage")
+
+        self.conn.set_timestamp(f'oldest_timestamp={self.timestamp_str(1)}')
+
+        # Create and populate a table.
+        uri = self.uri_prefix
+        self.session.create(uri, self.create_params)
+
+        self.populate(uri, self.table_numkv)
+
+        # Write to disk.
+        self.session.checkpoint()
+
+        # Remove 1/4 of the keys at timestamp 3.
+        c = self.session.open_cursor(uri, None)
+        for i in range(self.table_numkv // 4):
+            self.session.begin_transaction()
+            c.set_key(i)
+            c.remove()
+            self.session.commit_transaction(f'commit_timestamp={self.timestamp_str(3)}')
+        c.close()
+
+        self.session.checkpoint()
```

Comment:
I don't think this is needed as we are closing the connection which should perform a final checkpoint.